apps/aecore/src/aec_tx_pool.erl

```diff
@@ -236,37 +236,42 @@ int_get_max_nonce(Mempool, Sender) ->
 
 %% Ensure ordering of tx nonces in one account, and for duplicate account nonces
 %% we only get the one with higher fee or - if equal fee - higher gas price.
-int_get_candidate(#state{ db = Db, gc_db = GCDb }, MaxNumberOfTxs, BlockHash) ->
+int_get_candidate(#state{ db = Db, gc_db = GCDb }, MaxGas, BlockHash) ->
     {ok, Trees} = aec_chain:get_block_state(BlockHash),
     {ok, Header} = aec_chain:get_header(BlockHash),
-    int_get_candidate(MaxNumberOfTxs, {Db, GCDb}, ets:first(Db),
+    int_get_candidate(MaxGas, {Db, GCDb}, ets:first(Db),
                       {account_trees, aec_trees:accounts(Trees)},
                       aec_headers:height(Header), gb_trees:empty()).
 
-int_get_candidate(0, _Dbs, _, _AccountsTree, _Height, Acc) ->
-    gb_trees:values(Acc);
-int_get_candidate(_N, _Dbs, '$end_of_table', _AccountsTree, _Height, Acc) ->
-    gb_trees:values(Acc);
-int_get_candidate(N, Dbs = {Db, GCDb}, ?KEY(_, _, Account, Nonce, TxHash) = Key,
+int_get_candidate(Gas, Dbs = {Db, GCDb}, ?KEY(_, _, Account, Nonce, TxHash) = Key,
                   AccountsTree, Height, Acc) ->
     Next = ets:next(Db, Key),
     case gb_trees:is_defined({Account, Nonce}, Acc) of
         true ->
             %% The earlier must have had higher fee. Skip this tx.
-            int_get_candidate(N, Dbs, Next, AccountsTree, Height, Acc);
+            int_get_candidate(Gas, Dbs, Next, AccountsTree, Height, Acc);
         false ->
             Tx = ets:lookup_element(Db, Key, ?VALUE_POS),
-            TTL = aetx:ttl(aetx_sign:tx(Tx)),
-            case Height < TTL andalso ok == int_check_nonce(Tx, AccountsTree) of
+            Tx1 = aetx_sign:tx(Tx),
+            TxTTL = aetx:ttl(Tx1),
+            TxGas = aetx:gas(Tx1),
+            case Height < TxTTL andalso ok =:= int_check_nonce(Tx, AccountsTree) of
                 true ->
-                    NewAcc = gb_trees:insert({Account, Nonce}, Tx, Acc),
-                    int_get_candidate(N - 1, Dbs, Next, AccountsTree, Height, NewAcc);
+                    case Gas - TxGas of
+                        RemGas when RemGas >= 0 ->
+                            Acc1 = gb_trees:insert({Account, Nonce}, Tx, Acc),
+                            int_get_candidate(RemGas, Dbs, Next, AccountsTree, Height, Acc1);
+                        _ ->
+                            gb_trees:values(Acc)
```

Comment:
Just because this transaction didn't fit does not necessarily mean that no other transaction would?

Reply:
Yes, others could fit. It'd mean going through the whole pool until the one that fits is found or `$end_of_table` is reached. I'll fix it.